### PR TITLE
Fix for the flaky curl test

### PR DESCRIPTION
**Pull request by gearama**
See the comments in the code. 
The bottom line is that there is a race condition between EXPECT_EQ and the execution of the cleanup thread started by the curl session destruction which moves the connection to the pool then starts the cleanup thread. 

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/core/azure-core/test/ut/azure_libcurl_core_main_test.cpp [hunk 1]
```diff
@@ -53,11 +53,21 @@ namespace Azure { namespace Core { namespace Test {
       EXPECT_TRUE(session->m_keepAlive);
       EXPECT_FALSE(session->m_connectionUpgraded);
     }
+    // here the session is destroyed and the connection is moved to the pool
+    // the same destructor also makes a call to start the cleanup thread
+    // which will sleep for DefaultCleanerIntervalMilliseconds then loop through the connections
+    // in the pool and check for the ones that are expired(DefaultConnectionExpiredMilliseconds) and remove them
+    // which will be the case here if we wait long enough. 
+    // the test is flaky due to the fact that tests in the CI pipeline might take longer than 90 sec to execute
```
Comment (LarryOsterman) [inline sdk/core/azure-core/test/ut/azure_libcurl_core_main_test.cpp:61]: I might rephrase this as "without a delay, the test will be flaky" or something similar.

Also, I'm not sure that the new version tests the same thing as the old test did. The old test verifies that the connection is added to the cache after the session is destroyed, the new test verifies that the cache is cleared.

I think both checks are valid and important.

**Review by LarryOsterman — APPROVED** [review]

**Review by RickWinter — APPROVED** [review]

File: sdk/core/azure-core/test/ut/azure_libcurl_core_main_test.cpp [hunk 2]
```diff
@@ -52,12 +58,53 @@ namespace Azure { namespace Core { namespace Test {
...
+    // remove them which will be the case here if we wait long enough. the test is flaky due to the
+    // fact that tests in the CI pipeline might take longer than 90 sec to execute thus the cleanup
+    // thread strikes. to have this test be predictable we should let things run to completion and
+    // then check the pool size thus the sleep below plus another second to let the for loop in the
+    // thread do its thing.
+
+    auto t2 = high_resolution_clock::now();
+
+    /* Getting number of milliseconds as an integer. */
+    auto ms_int = duration_cast<milliseconds>(t2 - t1);
```
Comment (ahsonkhan) [inline sdk/core/azure-core/test/ut/azure_libcurl_core_main_test.cpp:76]: This local variable seems to be unsused. Why are we getting the ms both as an int and as a double?

File: sdk/core/azure-core/src/http/curl/curl_connection_pool_private.hpp [hunk 3]
```diff
@@ -29,102 +29,104 @@ namespace Azure { namespace Core { namespace Test {
...
+     * This pool offers static methods and it is allocated statically. There can be only one
+     * connection pool per application.
+     */
+    class CurlConnectionPool final {
 #if defined(TESTING_BUILD)
-    // Give access to private to this tests class
-    friend class Azure::Core::Test::CurlConnectionPool_connectionPoolTest_Test;
-    friend class Azure::Core::Test::CurlConnectionPool_uniquePort_Test;
-    friend class Azure::Core::Test::CurlConnectionPool_connectionClose_Test;
+      // Give access to private to this tests class
```
Comment (ahsonkhan) [inline sdk/core/azure-core/src/http/curl/curl_connection_pool_private.hpp:47]: There seems to be unnecessary whitespace changes in this file. Revert those?

File: sdk/core/azure-core/test/ut/azure_libcurl_core_main_test.cpp (hunk 2 above)
Comment (gearama) [reply sdk/core/azure-core/test/ut/azure_libcurl_core_main_test.cpp]: removed

File: sdk/core/azure-core/src/http/curl/curl_connection_pool_private.hpp (hunk 3 above)
Comment (gearama) [reply sdk/core/azure-core/src/http/curl/curl_connection_pool_private.hpp]: done, fro fun i used the clang-format-11 tool in wsl to format the file, the previous one was VS autoformatting the file

**Review by antkmsft — APPROVED** [review]
Thank you! Let's get this in!

**Comment by antkmsft** [comment]
@gearama, if there are no blockers to merge this, let's merge
